Firestore/core/src/firebase/firestore/auth/token.cc

```diff
@@ -22,6 +22,9 @@ namespace auth {
 
 Token::Token(const absl::string_view token, const User& user)
     : token_(token), user_(user), is_valid_(true) {
+  // We use "" to represent the absence of a token, which should
+  // be the case if the user is unauthenticated.
+  FIREBASE_ASSERT((!user_.is_authenticated()) == (token == ""));
```

Comment:
If this assertion failed I'm not sure what I would do about it; as it stands I can barely understand it.

Secondarily this seems like a weak assertion given that we're filling in is_valid_(true) above.

Tertiarily this seems backwards: we should be constructing the right kind token given our inputs rather than requiring every caller to get this right. We clearly haven't done that here so I propose we move all the callers to a factory method that returns a valid token given valid inputs or an Invalid token otherwise. At that point we don't have to fail if auth gives us a crazy value:

If the user is unauthenticated or the token is empty return invalid. Otherwise fill in a valid Token (and this constructor becomes private).

In the interest of good citizenry we could also make the assertion that these align, but that would be for reporting bad values to FIRAuth and less for catching our code out.